### PR TITLE
Favourites Bugfix #0038792: Configure Button in Favourites shows empty page

### PR DESCRIPTION
**Pull request by fhelfer**
https://mantis.ilias.de/view.php?id=38792

### CODE REVIEW

**Comment by alex40724** [comment]
@fhelfer Thanks for the PR. I am losing a bit overview regarding the dashboard changes. Could you point me to the feature request that specifies the modal behaviour, please? Thanks.

EDIT: Ok, found it. It's a dev issue here https://docu.ilias.de/goto_docu_wiki_wpage_7504_1357.html

File: Services/Repository/Favourites/classes/class.ilFavouritesListGUI.php
Comment (alex40724) [inline Services/Repository/Favourites/classes/class.ilFavouritesListGUI.php]: The implementation inside the modal is part of the Dashboard service (ilSelectedItemsBlockGUI). Please move these lines to the Dashboard service, too. E.g. the Dashboard could provide the complete modal instance.

File: Services/Repository/Favourites/templates/default/tpl.favourites.html
Comment (alex40724) [inline Services/Repository/Favourites/templates/default/tpl.favourites.html]: The only purpose of this template seems to be to integrate a js from the Dashboard service, please move this template to the Dashboard service, too.

File: Services/Repository/RecommendedContent/classes/class.ilDashboardRecommendedContentGUI.php
Comment (alex40724) [inline Services/Repository/RecommendedContent/classes/class.ilDashboardRecommendedContentGUI.php]: Is the PR mixing favourite and recommended content issues on purpose or is this a mistake?

**Review by alex40724 — CHANGES_REQUESTED** [review]
Thanks for the PR. I think we are all struggeling with the borders of these services.

Please check my comments, may be @iszmais can assist here. I do not want to make it too complicated, but also not mix up things even more.

If the suggestions are feasible, please implement them. If not, please let me know.

Best regards
Alex

**Comment by fhelfer** [comment]
> fhelfer Thanks for the PR. I am losing a bit overview regarding the dashboard changes. Could you point me to the feature request that specifies the modal behaviour, please? Thanks.
> 
> EDIT: Ok, found it. It's a dev issue here https://docu.ilias.de/goto_docu_wiki_wpage_7504_1357.html

Exactly thats the Feature request. Unfortunately a few buttons still point to dashbaord despite the user being in any other contexts. 
A better approach to handling favorite items will be targeted in later FW articles, targeting also dashboard blocks in general. Also sortations in the modal will be my next PR :)

> Thanks for the PR. I think we are all struggeling with the borders of these services.
>Please check my comments, may be iszmais can assist here. I do not want to make it too complicated, but also not mix up things even more.
>If the suggestions are feasible, please implement them. If not, please let me know.

Unfortunately i cannot implement your changes until after my vacation. But will look after this asap after.

Best, @fhelfer

File: Services/Repository/Favourites/classes/class.ilFavouritesListGUI.php
Comment (fhelfer) [reply Services/Repository/Favourites/classes/class.ilFavouritesListGUI.php]: Hmm, this is very specific code to FavouritesList. The AsyncRenderURL and onload are different to what DashboardBlocks need.
I dont think this is a pretty solution. Not much code would be moved.

File: Services/Repository/Favourites/templates/default/tpl.favourites.html
Comment (fhelfer) [reply Services/Repository/Favourites/templates/default/tpl.favourites.html]: This is also specifically needed for favourites in slate. DashboardBlock Modals do not want to send js contents on every modal render.

File: Services/Repository/RecommendedContent/classes/class.ilDashboardRecommendedContentGUI.php
Comment (fhelfer) [reply Services/Repository/RecommendedContent/classes/class.ilDashboardRecommendedContentGUI.php]: removed

File: Services/Repository/Favourites/templates/default/tpl.favourites.html
Comment (alex40724) [reply Services/Repository/Favourites/templates/default/tpl.favourites.html]: The repository component should not be responsible to load dashboard javascript code. This violates SoC principles. This results in high coupling and bad maintainability. As I wrote I understand that there is currently no clear interface, but these lines may produce severe issues in the future. A Dashboard maintainer should be able to change or even delete this JS file without the need to check other components.

File: Services/Repository/Favourites/classes/class.ilFavouritesListGUI.php
Comment (alex40724) [reply Services/Repository/Favourites/classes/class.ilFavouritesListGUI.php]: It is not about the amount of code. It is about separation of concerns and coupling.

The Dashboard service provides the management view, since ilPDSelectedItemsBlockGUI is a part of it. This is already Favourites specific code. On the other hand the knowledge of things like the ctrl path ilDashboardGUI::class, ilColumnGUI::class, ilSelectedItemsBlockGUI::class are Dashboard internals. Providing only the modal instance would reduce the coupling and keep these Dashboard internals out of the Repository service.

**Comment by iszmais** [comment]
Hi @alex40724 ,

i see this more as a conceptional issue (wich already made issues within the past) than a technical one. I try to summarize my general concern:
ATM the have a not fully abstracted concept of favorites. The functionality, location and naming of the favorites imply that they are exclusively part of the dashboard functionality (e.g. the command for add Favorites is "addToPersonalDesktop" aka Dashboard), but the handling within the menu implies otherwise that the favorites are an abstract component. This state proposes 2 solutions which can be considered best practice IMO:
- Case 1 "Favorites are an independent functionality": This would require the abstraction of said modal into an independent space in ILIAS. However this would also imply a full refactoring of the functionality since, even separated into a different component, the dependency to the dashboard may be reduced but cannot be fully eradicated.
- Case 2 "Favorites are a dashboard functionality": In this case the menu entry should not interfere with the configuration process of the dashboard and therefore the "Configure" link should be removed. It could be replaced with a link to the dashboard on wich the user than can configure his favorites (Similar to the current handling of mail, tags, .. inside the mainmenu and metabar). The menu would only provide a readonly view of said component just like for the repository tree or last visited.

I would highly favor this last approach since it unifies the menu entry with the convention of other menu entries and reduces another dual configuration and therefore the complexity of configruation within ILIAS in general.

Pls let me know if you have any concerns about that conceptional change.
If not i would prepare a proposal to take this to the Jour Fixe.

Greetings,
@iszmais

**Comment by alex40724** [comment]
Hi @iszmais,

thanks for the answer.

My understanding was, that the Dashboard component aims to provide an interface that other components (like favourties) should serve in the future. This it at least stated in the current roadmap of the Dashboard: "In detail a component should provide all the necessary information on its presentation and required settings to the `Dashboard` by implementing a `Provider Interface`."

I am aware that we don't have an implementation for this now, but it would be great to have at least an outline of this future interface. This would make it easier to come to conclusions for these kind of discussions now imo.

You are writing "ATM the have a not fully abstracted concept of favorites.".

If the favorites component serves an interface of the dashboard, we would not need that imo.

We can decide on this in the Jour Fixe. Currently the favorites core classes are located in the Repository service. It would be ok for me to move these somewhere else.

Best

**Comment by matthiaskunkel** [comment]
Jour Fixe, 15 APR 2024: We had a short discussion about the future of the Favourites service and accept the suggestion by @fhelfer and @alex40724 to move the related classes to the Dashboard component and make the function Favourites a part of the Dashboard. This change is accepted for 9 and trunk.

File: lang/ilias_de.lang
Comment (nicoroeser) [inline lang/ilias_de.lang]: Nitpick: „zum entfernen“ → „zum Entfernen“ (capitalisation).

**Comment by iszmais** [comment]
@fhelfer Can you move the Favourites-classes to the Dashboard?
I would recommend the same structure as in Repository (simple subdirectory inside classes).

Thanks alot!
@iszmais

**Comment by fhelfer** [comment]
@iszmais Code has been moved and configure Modal is working atm.
Please give the PR another review and also port changes to trunk after merge

Thanks, @fhelfer

**Comment by iszmais** [comment]
@fhelfer Works fine thanks a lot for your collaboration!

Greetings,
@iszmais

**Pull request merged**
